Add trimming annotations to DotNetObjectReference

eerhardt opened the pull request:
DotNetDispatcher has some unbounded reflection against DotNetObjectReference Types that the trimmer can't understand. Adding trimming annotations to tell the trimmer to preserve the public methods on the TValue of DotNetObjectReference instances, so the JSInvokable methods won't be trimmed.

See https://github.com/dotnet/maui/issues/6965

cc @vitek-karas

eerhardt committed:
Add trimming annotations to DotNetObjectReference

DotNetDispatcher has some unbounded reflection against DotNetObjectReference Types that the trimmer can't understand. Adding trimming annotations to tell the trimmer to preserve the public methods on the TValue of DotNetObjectReference instances, so the JSInvokable methods won't be trimmed.

See https://github.com/dotnet/maui/issues/6965

eerhardt committed:
Fix build

eerhardt committed:
Resolve ILLink warnings and update the warning baselines.

eerhardt committed:
Use LinkerFlags for better readability.

vitek-karas commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -381,9 +381,11 @@ private static Task GetTaskByType(Type type, object obj)
     private static (MethodInfo methodInfo, Type[] parameterTypes) GetCachedMethodInfo(IDotNetObjectReference objectReference, string methodIdentifier)
     {
-        var type = objectReference.Value.GetType();
+        var type = objectReference.Type;
```
There is a difference here potentially.
```C#
class MyValueBase {}
class MyDerivedValue : MyValueBase {  void ImportantMethod() {} }

void Test()
{
    DotnetObjectReference.Create<MyValueBase>(new MyDerivedValue());
}
```

Before this change the code would build the cache with the `ImportantMethod` in it. After this change the `ImportantMethod` will not be there and thus won't be callable from JS.

We could add an exception/assert into the .ctor of `DotNetObjectReference` that the `value.GetType() == typeof(TValue)`.

eerhardt replied:
I think this is an acceptable change, in order to make this path trimmable (which is needed for both wasm and Maui).

> We could add an exception/assert into the .ctor of DotNetObjectReference that the value.GetType() == typeof(TValue).

This would have to be an exception, since the API is public.

eerhardt replied:
@SteveSandersonMS do you have an opinion here? Would it be beneficial to add an exception check to `DotNetObjectReference<T>`'s ctor to ensure the Types match?

javiercn replied:
I think this might be problematic as it might prevent some patterns where the type used is a base class (can't remember if we can use interfaces here).

eerhardt replied:
Can you point me to them? `DotNetObjectReference<TValue>`'s ctor is internal, and it's only caller is `DotNetObjectReference.Create<TValue>(TValue value)`.

According to https://source.dot.net/#Microsoft.JSInterop/DotNetObjectReference.cs,90b7156b4279e601,references, there are only 3 callers to the Create method, all which don't use base classes.

eerhardt replied:
OK, I can change this back to `var type = objectReference.Value.GetType();`. The annotations I'm adding here will allow the "normal" case to work successfully in a trimmed app.

We will still have a trim warning here, which we can add to the baseline. It was previously being suppressed in code, which isn't valid for warnings that are real trimming problems. To address that warning in a future change, maybe we can add a new `DotNetObjectReference.CreateExact` (or something with a better name) that is guaranteed to be safe for trimming. Then the existing `DotNetObjectReference.Create` method can be annotated with `RequiresUnreferencedCode`, so any callers will get trim warnings. And they can switch over to the new, trim-compatible API.

Thoughts on this plan?

eerhardt replied:
I've updated the PR to follow the above proposal. Please take a look and let me know what you think.

MackinnonBuck commented:
Related: #38044

JamesNK commented:
FYI the PR that re-enables trimming analysis on the solution found a number of trimming warnings in JSInterop. Not sure if related

https://github.com/dotnet/aspnetcore/pull/41683

eerhardt commented:
Looking through the warnings on https://github.com/dotnet/aspnetcore/pull/41683#issuecomment-1126617181, those are different warnings. This one was being suppressed on the whole DotNetDispatcher class. I am removing the suppression with this PR.

eerhardt committed:
Respond to PR feedback.

Remove the breaking change to only allow methods on the TValue type of IDotNetObjectReference. If a caller provides a base class to the TValue of DotNetObjectReference, and a derived type as the actual instance, only the base class's methods will be preserved.

This allows for the current behavior to be preserved, at the cost of not being fully trim-compatible. In order to make it fully trim-compatible, we will need to introduce a new API that can be made trimming compatible.

vitek-karas commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -499,11 +499,16 @@ private static Assembly GetRequiredLoadedAssembly(AssemblyKey assemblyKey)
-        _cachedConvertToTaskByType.Clear();
+        public static void ClearCache(Type[]? _)
+        {
+            _cachedMethodsByAssembly.Clear();
+            _cachedMethodsByType.Clear();
+            _cachedConvertToTaskByType.Clear();
```
Unrelated to this change - but that's where I noticed this. These caches will effectively disable support for unloadable (collectible) assemblies - since they hold a string reference to the `Type`. Would be really nice if we didn't introduce more hurdles in this space in the new frameworks.

javiercn replied:
@vitek-karas that's fair, but at the same time its a trade-off between perf and `correctness` on a very concrete scenario. Do we have a design/proposal on how to deal with these two concerns that we could apply across all frameworks?

In any case, that's something we should discuss outside of this PR, but if we come up with a reasonable design for it, I think we would incorporate it.

javiercn reviewed (approved):
Looks great!

eerhardt commented:
Any more feedback here? Or is this good to merge?

javiercn commented:
@eerhardt I think it is good to merge. To be fair, I am not an expert in the trimming area, but as far as I understand, the changes make sense.
